fuse/nodefs/fsconnector.go

```diff
@@ -126,9 +135,23 @@ func (c *FileSystemConnector) forgetUpdate(nodeID uint64, forgetCount int) {
 
 	if forgotten, handled := c.inodeMap.Forget(nodeID, forgetCount); forgotten {
 		node := (*Inode)(unsafe.Pointer(handled))
+		if node.IsDir() && len(node.Children()) > 0 {
+			// This directory node has children that have not yet been forgotten.
+			// We don't want to remove it from the tree (i.e. remove it from its parent)
+			// because the children may have parent pointers to it.
+			// A later LOOKUP will return this node with a new NodeID.
+			return
+		}
 		node.mount.treeLock.Lock()
-		c.recursiveConsiderDropInode(node)
+		// Remove Inode from all parents
```

Comment:
Great. I didn't do a map at the time because I thought it would be
expensive, but TBH I did not measure it. If it is an issue, maybe you can
have a single parent pointer in case of directory nodes, eg.

  // use in case this is a directory
  dirParent *pathNode
  // use for non-directory nodes.
  parents  map[*pathNode]struct{}



On Wed, Mar 23, 2016 at 10:56 AM rfjakob <notifications@github.com> wrote:

> In fuse/nodefs/fsconnector.go
> <https://github.com/hanwen/go-fuse/pull/103#discussion_r57132915>:
>
> >  		node.mount.treeLock.Lock()
> > -		c.recursiveConsiderDropInode(node)
> > +		// Remove Inode from all parents
>
> Yes, there already is a "Parent" pointer - this needs to become a map. I
> will rework the patch series and start a new pull request once I am ready!
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/hanwen/go-fuse/pull/103/files/004bb8a47f407fea15ba2caedcb216371ca806f4#r57132915>
>